builtin/logical/transit/path_decrypt.go

```diff
@@ -44,64 +97,143 @@ func (b *backend) pathDecrypt() *framework.Path {
 
 func (b *backend) pathDecryptWrite(
 	req *logical.Request, d *framework.FieldData) (*logical.Response, error) {
-	name := d.Get("name").(string)
-	ciphertext := d.Get("ciphertext").(string)
-	if len(ciphertext) == 0 {
-		return logical.ErrorResponse("missing ciphertext to decrypt"), logical.ErrInvalidRequest
+
+	// Get the policy
+	p, lock, err := b.lm.GetPolicyShared(req.Storage, d.Get("name").(string))
+	if lock != nil {
+		defer lock.RUnlock()
+	}
+	if err != nil {
+		return nil, err
+	}
+	if p == nil {
+		return logical.ErrorResponse("policy not found"), logical.ErrInvalidRequest
 	}
 
-	var err error
+	batchInputRaw := d.Get("batch").(string)
+	var batchInput []byte
+	if len(batchInputRaw) != 0 {
+		batchInput, err = base64.StdEncoding.DecodeString(batchInputRaw)
+		if err != nil {
+			return logical.ErrorResponse("failed to base64-decode batch input"), logical.ErrInvalidRequest
+		}
+	} else {
+		ciphertext := d.Get("ciphertext").(string)
+		if len(ciphertext) == 0 {
+			return logical.ErrorResponse("missing ciphertext to decrypt"), logical.ErrInvalidRequest
+		}
 
-	// Decode the context if any
-	contextRaw := d.Get("context").(string)
-	var context []byte
-	if len(contextRaw) != 0 {
-		context, err = base64.StdEncoding.DecodeString(contextRaw)
+		var singleItemBatch []BatchDecryptionItemRequest
+		singleItemBatch = append(singleItemBatch, BatchDecryptionItemRequest{
+			Ciphertext: ciphertext,
+			Context:    d.Get("context").(string),
+			Nonce:      d.Get("nonce").(string),
+		})
+
+		batchInput, err = jsonutil.EncodeJSON(singleItemBatch)
```

Comment:
This is super duper inefficient, and we want `transit` to be speedy. You should do the opposite of this: if there is batch input, decode it to an array of items and append that to the final array. Otherwise, append a single request that is made up of the request data. You should also make the array manually with a capacity of 1 so that in the common, non-batch case it's a single memory allocation with no resize.